JDK-8289332: Auto-generate ids for user-defined headings

hns opened the pull request:
This is a simple change to automatically generate `id` attributes for HTML headers in documentation comments. 

I decided to leave the functionality in `HtmlDocletWriter` rather than moving it to `HtmlIds` because it uses the same helper methods as`commentTagsToContent`.

The value of the `id` attribute id derived from the content of the header tag with spaces and other non-word characters replaced by `-`. A `-hdr` suffix is appended to avoid collisions with other ids. If there are duplicate values an int counter is appended to make them unique.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8289332](https://bugs.openjdk.org/browse/JDK-8289332): Auto-generate ids for user-defined headings


### Reviewers
 * [Jonathan Gibbons](https://openjdk.org/census#jjg) (@jonathan-gibbons - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9627/head:pull/9627` \
`$ git checkout pull/9627`

Update a local copy of the PR: \
`$ git checkout pull/9627` \
`$ git pull https://git.openjdk.org/jdk pull/9627/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9627`

View PR using the GUI difftool: \
`$ git pr show -t 9627`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9627.diff">https://git.openjdk.org/jdk/pull/9627.diff</a>

</details>

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java:
```diff
@@ -1752,6 +1753,53 @@ protected Boolean defaultAction(DocTree node, Content content) {
+                return; // Avoid generating id if embedded <a id=...> is present
+            }
+        }
+        String idValue = sb.toString()
+                .toLowerCase(Locale.ROOT)
+                .trim()
+                .replaceAll("[^\\w_-]+", "-");
+        if (!idValue.isEmpty()) {
+            // Make id unique
+            idValue = idValue + "-hdr";
+            if (!headerIds.add(idValue)) {
+                int counter = 1;
+                while (!headerIds.add(idValue + counter)) {
+                    counter++;
+                }
+                idValue = idValue + counter;
+            }
+            content.add("id=\"").add(idValue).add("\"");
+        }
```
This part of the algorithm, to create an `id` from the text contents of a header, still feels like it belongs in `HtmlIds`.  It feels like we should have 
```
HtmlIds.getIdForHeader(CharSequence headerText)
```
or something like that

hns replied:
I moved the code to a new `HtmlIds.forHeading(CharSequence, Set<String>)` method. The `forHeading` follows the naming scheme of `HtmlIds`. The second argument of type `Set<String>` is to make the returned id unique within its page.

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java:
```diff
@@ -1752,6 +1753,53 @@ protected Boolean defaultAction(DocTree node, Content content) {
+                .replaceAll("[^\\w_-]+", "-");
+        if (!idValue.isEmpty()) {
+            // Make id unique
+            idValue = idValue + "-hdr";
```
Would it be better to use `header` rather than just `hdr` ?
Better still, `heading` seems to be the term in the HTML spec; `header` is also used, but not for headings.

hns replied:
I renamed the `-hdr` suffix to `-heading` and also renamed code occurrences of "header" to "heading" (except for the test name).

jonathan-gibbons reviewed (commented):
Some inline comments/questions.

hns commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java:
```diff
@@ -1752,6 +1753,53 @@ protected Boolean defaultAction(DocTree node, Content content) {
+
+    private boolean hasIdAttribute(StartElementTree node) {
+        return node.getAttributes().stream().anyMatch(
+                dt -> equalsIgnoreCase(((AttributeTree)dt).getName(), "id"));
```
Cood catch, I added a pattern matching instanceof which should be the easiest way to fix the problem.

hns commented:
Thanks for the review, I think I addressed all your feedback in the incremental commit.

As a corner case I removed the emptiness check for the heading text, so an empty heading will now get an `id="-heading[idx]"` attribute which is useless but also harmless.

jonathan-gibbons reviewed (approved):
yes, I like the change to move the code to allocate the `id` to `HtmlIds`
